azure.yaml

```diff
@@ -1 +1,33 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/Azure/azure-dev/azure-dev-cli_1.0.1/schemas/v1.0/azure.yaml.json
 name: azcrypto
+requiredVersions:
+  azd: '>= 1.0.1'
+hooks:
+  # cspell:ignore postprovision,newkey,keyout
+  postprovision:
+    shell: sh
+    run: |
+      if [ "${AZURE_MANAGEDHSM}" = true ] && ! [ -e "./infra/out/${AZURE_ENV_NAME}/${AZURE_KEYVAULT_NAME}.json" ]; then
+        # Activate Managed HSM.
+        mkdir -p "./infra/out/${AZURE_ENV_NAME}/"
+
+        openssl req -newkey rsa:2048 -subj "/CN=${AZURE_KEYVAULT_NAME}/" -nodes -keyout "./infra/out/${AZURE_ENV_NAME}/cert_0.key" -x509 -days 30 -out "./infra/out/${AZURE_ENV_NAME}/cert_0.cer"
+        openssl req -newkey rsa:2048 -subj "/CN=${AZURE_KEYVAULT_NAME}/" -nodes -keyout "./infra/out/${AZURE_ENV_NAME}/cert_1.key" -x509 -days 30 -out "./infra/out/${AZURE_ENV_NAME}/cert_1.cer"
+        openssl req -newkey rsa:2048 -subj "/CN=${AZURE_KEYVAULT_NAME}/" -nodes -keyout "./infra/out/${AZURE_ENV_NAME}/cert_2.key" -x509 -days 30 -out "./infra/out/${AZURE_ENV_NAME}/cert_2.cer"
+        az keyvault security-domain download --hsm-name "${AZURE_KEYVAULT_NAME}" --sd-wrapping-keys "./infra/out/${AZURE_ENV_NAME}/cert_0.cer" "./infra/out/${AZURE_ENV_NAME}/cert_1.cer" "./infra/out/${AZURE_ENV_NAME}/cert_2.cer" --sd-quorum 2 --security-domain-file "./infra/out/${AZURE_ENV_NAME}/${AZURE_KEYVAULT_NAME}.json"
+
+        # Add principal as Managed HSM Crypto User.
+        az keyvault role assignment create --hsm-name "${AZURE_KEYVAULT_NAME}" --role "21dbd100-6940-42c2-9190-5d6cb909625b" --scope "/" --assignee-object-id "${AZURE_PRINCIPAL_ID}"
+
+        # Wait 30s and create keys via data plane operations.
+        sleep 30s
+
+        # TODO: Consider re-invoking `azd provision` with an env var / param to provision with keys.
```

Comment:
@ellismg @weikanglim I was curious about your thoughts on "recursive azd provisioning". This is a case where it's impossible to completely provision a resource (rather, nested resources) because manual steps must be performed first: above, we create several certs meant to be distributed and then use them to activate a Managed HSM. You can't do anything with the Managed HSM (similar to Key Vault, but even more secure) - not create keys, not manage data-plane RBAC - until it's activated. But I'd like to keep key provisioning in one place, even if that itself was a bicep module. What do you think about invoking `azd provision` again with env vars (or arguments, which I plan to add soon) to say "now this time, also provision the keys"?

Alternatively, I suppose I could still use a separate bicep module with necessary scope and conditions and then just `az deployment group create` it specifically, but then I have to also worry about authenticating `az` and not just `azd`. Or, could I auth with `az` and `azd` could use `azidentity.NewDefaultAzureCredential` that should pick it up? I noticed it doesn't seem to do that currently.